### ISSUE TITLE
ServiceWorker's Fetch Handler Doesn't Return a Valid Response

### ISSUE THREAD

**Issue by JammSpread** [coder/code-server#2760]
Around mid-2021 Google seems to want to enforce PWAs have proper fetch handlers.
The consequence being no installability on at least Chrome.

I think it would be best to make sure that this is resolved prior to that "deadline".
(Although it's February so this may be *too* early.)

I'm thinking either #2007 (which may require that #1925 be reopened) could be reverted, or a page similar to the error page could be returned notifying the user that `code-server` cannot be used offline.

(You may be able to test this functionality by using the Chrome flag `check-offline-capability`.)

See more:
- https://web.dev/install-criteria/#criteria
- https://developer.chrome.com/blog/improved-pwa-offline-detection/

**Comment by JammSpread** [comment]
> I'm thinking either #2007 (which may require that #1925 be reopened) could be reverted, or a page similar to the error page could be returned notifying the user that code-server cannot be used offline.

(Maybe the latter may look something like this):

![Offline](https://user-images.githubusercontent.com/61063879/108952755-18b9ad80-7638-11eb-8842-54c85935ead2.png)

**Comment by jsjoeio** [comment]
> The consequence being no installability on at least Chrome.

We definitely don't want that to happen. Thanks for bringing this up sooner rather than later 🙌

I did see this as well in the link you provided:
> TL;DR: Offline support has been part of the PWA installability criteria since the beginning. We are updating the offline detection logic to ensure a PWA actually provides an offline experience, closing the loophole used by some developers to meet the installability critiera. In the future, **sites with empty fetch event handlers will no longer meet the criteria**.

I'm actually in the middle of adding tests for our service worker code and thought it was strange we don't actually use the `fetch` handler in our service worker. It does nothing and I even have a test to make sure it does nothing 😂  (see [here](https://github.com/cdr/code-server/blob/f1f0f497ea4aad6e412af93863aeeeadd6378e83/test/serviceWorker.test.ts#L77-L81)).

^I like the page you designed here above. Matches our other pages and communicates the issue clearly.

It sounds like we'll need to do two things:
- [ ] fix serviceWorker fetch handler (see [example](https://web.dev/offline-fallback-page/#the-service-worker-code))
- [ ] add offline fallback page (see [examples](https://web.dev/offline-fallback-page/#the-service-worker-code))

**Comment by jsjoeio** [comment]
Adding `help-wanted` in case anyone wants to grab this. Otherwise, I'll probably end up doing it

**Comment by vapurrmaid** [comment]
@jsjoeio Could be a great one for me to pick up

**Comment by jsjoeio** [comment]
Woo! All you! ♥️

**Comment by vapurrmaid** [comment]
@jsjoeio - I'm promising this by the first June milestone, with optimal hopes to get to it in the last week of May. Feel free to modify the milestones around this - and thanks for the patience!

**Comment by vapurrmaid** [comment]
Interesting updated - from one of the linked to articles [Improving Progressive Web App offline support detection](https://developer.chrome.com/blog/improved-pwa-offline-detection/)

An edit has been made:

> Updated April 14th, 2021: We previously announced plans to update the installability criteria to ensure a PWA actually provides an offline experience. After listening to your feedback, and discovering a number of issues, we have decided to put those plans on hold. We strongly believe providing a valid page when the user is offline is critical to providing a good user experience.

---

It seems we have more time to implement #2760 , though I don't think that means we should hold off. What it does mean, however, is we must continually monitor emerging PWA technologies.

**Comment by jsjoeio** [comment]
Not stale

**Comment by douglasparker** [comment]
@code-asher can you please communicate why this is closed as not planned, so everyone following this issue isn’t left in the dark?

**Comment by code-asher** [comment]
It could be I misunderstood.  From what I could tell, this is not going to change, so there is nothing we have to implement?

**Comment by douglasparker** [comment]
> It could be I misunderstood. From what I could tell, this is not going to change, so there is nothing we have to implement?

Nothing that I have seen indicates that this is not going to change- Google has merely said that it is currently on hold.

**Comment by code-asher** [comment]
Definitely happy to reopen if that changes.  PRs also welcome!  Although, I think the long-term move might be to use VS Code's service worker rather than our own, but this will require some investigation.
